File: backend/modules/projects/service.py
from sqlalchemy.ext.asyncio import AsyncSession
from sqlalchemy.future import select
from fastapi import Depends, Request
from database import Project, get_db
from utils import APIException, get_user_from_username
from .schemas import (
    CreateProjectRequest,
    UpdateProjectRequest,
    PaginatedProjectResponse,
)
import logging

logging.basicConfig(
    level=logging.INFO, format="%(asctime)s - %(name)s - %(levelname)s - %(message)s"
)
logger = logging.getLogger(__name__)


class ProjectService:
    # Create Project
    async def create_project(
        self,
        user_id: str,
        project_request: CreateProjectRequest,
        db: AsyncSession = Depends(get_db),
    ) -> str:
        new_project = Project(user_id=user_id, **project_request.model_dump())
        try:
            db.add(new_project)
            await db.commit()
            await db.refresh(new_project)
        except Exception as e:
            await db.rollback()
            logger.error(f"Database error during project creation: {e}")
            raise APIException(
                message="An error occurred while creating project",
                status=500,
                error_code="SERVER_ERROR",
            )
        return "Project created successfully"

    # Update Project
    async def update_project(
        self,
        user_id: str,
        project_id: str,
        project_request: UpdateProjectRequest,
        db: AsyncSession = Depends(get_db),
    ) -> str:
        query = select(Project).where(Project.id == project_id)
        result = await db.execute(query)
        existing_project = result.scalars().first()

        if not existing_project:
            raise APIException(
                message="Project not found",
                status=404,
                error_code="PROJECT_NOT_FOUND",
            )

        for field, value in project_request.model_dump().items():
            if value is not None:
                setattr(existing_project, field, value)

        try:
            await db.commit()
            await db.refresh(existing_project)
        except Exception as e:
            await db.rollback()
            logger.error(f"Database error during project update: {e}")
            raise APIException(
                message="An error occurred while updating project",
                status=500,
                error_code="SERVER_ERROR",
            )
        return "Project updated successfully"

    # Get Project
    async def get_projects(
        self,
        username: str,
        page: int = 1,
        limit: int = 10,
        db: AsyncSession = Depends(get_db),
    ) -> PaginatedProjectResponse:
        try:
            user_id = await get_user_from_username(username, db)
            offset = (page - 1) * limit

            query = (
                select(Project)
                .where(Project.user_id == user_id)
                .order_by(Project.created_at.desc())
                .offset(offset)
                .limit(limit)
            )

            result = await db.execute(query)
            projects = result.scalars().all()

            return PaginatedProjectResponse(
                items=projects,
                page=page,
                limit=limit,
            )
        except Exception as e:
            logger.error(f"Database error during project fetch: {e}")
            raise APIException(
                message="An error occurred while fetching projects",
                status=500,
                error_code="SERVER_ERROR",
            )

        # Delete Project
        async def delete_project(
            self,
            user_id: str,
            project_id: str,
            db: AsyncSession = Depends(get_db),
        ) -> str:
            query = select(Project).where(Project.id == project_id)
            result = await db.execute(query)
            existing_project = result.scalars().first()

            if not existing_project:
                raise APIException(
                    message="Project not found",
                    status=404,
                    error_code="PROJECT_NOT_FOUND",
                )

            try:
                await db.delete(existing_project)
                await db.commit()
            except Exception as e:
                await db.rollback()
                logger.error(f"Database error during project deletion: {e}")
                raise APIException(
                    message="An error occurred while deleting project",
                    status=500,
                    error_code="SERVER_ERROR",
                )
            return "Project deleted successfully"

File: frontend/src/services/projects/service.ts
import { APIResponse } from "@/src/utils/apiResponse";
import {
  CreateProjectRequest,
  ProjectResponse,
  UpdateProjectRequest,
} from "./schema";
import { axios, logger } from "@/src/utils";

class ProjectService {
  async createProject(
    request: CreateProjectRequest
  ): Promise<APIResponse<null>> {
    try {
      logger.info("Create project flow started...");
      const response = await axios.post("/projects/create", request);
      logger.info("Create project successful!");
      return response.data;
    } catch (error) {
      logger.error("Create project failed :: error :: ", error);
      throw error;
    }
  }

  async updateProject(
    request: UpdateProjectRequest,
    project_id: string
  ): Promise<APIResponse<null>> {
    try {
      logger.info("Update project flow started...");
      const response = await axios.patch(
        `/projects/update/${project_id}`,
        request
      );
      logger.info("Update project successful!");
      return response.data;
    } catch (error) {
      logger.error("Update project failed :: error :: ", error);
      throw error;
    }
  }

  async getCurrentUserProjects(
    page: number = 1,
    limit: number = 10
  ): Promise<ProjectResponse[]> {
    try {
      logger.info("Get current user projects flow started...");
      const response = await axios.get(`/projects`, {
        params: {
          page: page,
          limit: limit,
        },
      });
      logger.info("Get current user projects successful!");
      return response.data.data.items;
    } catch (error) {
      logger.error("Get current user projects failed :: error :: ", error);
      throw error;
    }
  }

  async deleteProject(project_id: string): Promise<APIResponse<null>> {
    try {
      logger.info("Delete project flow started...");
      const response = await axios.delete(`/projects/delete/${project_id}`);
      logger.info("Delete project successful!");
      return response.data;
    } catch (error) {
      logger.error("Delete project failed :: error :: ", error);
      throw error;
    }
  }

  // PUBLIC ENDPOINT
  async getUserProjects(
    username: string,
    page: number = 1,
    limit: number = 10
  ): Promise<ProjectResponse[]> {
    try {
      logger.info("Get user projects flow started...");
      const response = await axios.get(`/projects/${username}`, {
        params: {
          page: page,
          limit: limit,
        },
      });
      logger.info("Get user projects successful!");
      return response.data.data.items;
    } catch (error) {
      logger.error("Get user projects failed :: error :: ", error);
      throw error;
    }
  }
}

const projectService = new ProjectService();
export default projectService;
